OrbitService/OrbitService.cpp

```diff
@@ -1,16 +1,74 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
 #include "OrbitService.h"
 
+#include <absl/strings/match.h>
+#include <fcntl.h>
+
+#include <chrono>
+#include <cstdio>
+
 #include "OrbitAsioServer.h"
 #include "OrbitGrpcServer.h"
 
+std::string ReadStdIn() {
+  char tmp = fgetc(stdin);
+  if (tmp == -1) return "";
+
+  std::string result;
+  do {
+    result += tmp;
+    tmp = fgetc(stdin);
+  } while (tmp != -1);
+
+  return result;
+}
+
+bool IsSshConnectionAlive(
+    std::chrono::time_point<std::chrono::steady_clock> last_ssh_message,
+    const int timeout_in_seconds) {
+  return std::chrono::duration_cast<std::chrono::seconds>(
+             std::chrono::steady_clock::now() - last_ssh_message)
+             .count() < timeout_in_seconds;
+}
+
 void OrbitService::Run(std::atomic<bool>* exit_requested) {
   std::cout << "Starting GRPC server at " << grpc_address_ << std::endl;
   std::unique_ptr<OrbitGrpcServer> grpc_server;
   grpc_server = OrbitGrpcServer::Create(grpc_address_);
 
   std::cout << "Starting Asio server on port " << asio_port_ << std::endl;
   OrbitAsioServer asio_server{asio_port_, tracing_options_};
-  asio_server.Run(exit_requested);
+
+  // make stdin non blocking
+  fcntl(STDIN_FILENO, F_SETFL, O_NONBLOCK);
+
+  // Main loop
+  while (!(*exit_requested)) {
+    asio_server.LoopTick();
+
+    std::string stdin_data = ReadStdIn();
+    // if ssh sends EOF, end main loop
+    if (feof(stdin)) break;
+
+    if (!stdin_data.empty()) {
+      last_stdin_message_ = std::chrono::steady_clock::now();
+
+      if (!ssh_watchdog_active_ &&
+          absl::StrContains(stdin_data, start_passphrase_)) {
+        ssh_watchdog_active_ = true;
+      }
+    }
```

Comment:
Nit: Feels a little bit unintuitive. Maybe you can branch first on whether the watchdog is enabled.

Also Nit: If you want to condense state you can combine last_stdin_message_ and ssh_watchdog_active_ into a std::optional<timepoint>.

Reply:
I reordered it a bit. I think it is more intuitive now